python/cuml/dask/ensemble/base.py

```diff
@@ -0,0 +1,179 @@
+                    model[worker],
+                    worker_data,
+                    convert_dtype,
+                    workers=[worker],
+                    pure=False)
+            )
+        wait_and_raise_from_futures(futures)
+        return self
+
+    def _concat_treelite_models(self):
+        """
+        Convert the cuML Random Forest model present in different workers to
+        the treelite format and then concatenate the different treelite models
+        to create a single model. The concatenated model is then converted to
+        bytes format.
+        """
+        model_protobuf_futures = list()
+        for w in self.workers:
+            model_protobuf_futures.append(
+                dask.delayed(_get_protobuf_bytes)
+                (self.rfs[w]))
+        mod_bytes = self.client.compute(model_protobuf_futures, sync=True)
+        last_worker = w
+        all_tl_mod_handles = []
+        model = self.rfs[last_worker].result()
+        all_tl_mod_handles = [model._tl_model_handles(pbuf_bytes)
+                              for pbuf_bytes in mod_bytes]
+
+        model._concatenate_treelite_handle(
+            treelite_handle=all_tl_mod_handles)
+        return model
+
+    def _predict_using_fil(self, X, delayed, **kwargs):
+        data = DistributedDataHandler.create(X, client=self.client)
+        self.datatype = data.datatype
+        return self._predict(X, delayed=delayed, **kwargs)
+
+    def _get_params(self, deep):
+        model_params = list()
+        for idx, worker in enumerate(self.workers):
+            model_params.append(
+                self.client.submit(
+                    _func_get_params,
+                    self.rfs[worker],
+                    deep,
+                    workers=[worker]
+                )
+            )
+        params_of_each_model = self.client.gather(model_params, errors="raise")
+        return params_of_each_model
+
+    def _set_params(self, **params):
+        model_params = list()
+        for idx, worker in enumerate(self.workers):
+            model_params.append(
+                self.client.submit(
+                    _func_set_params,
+                    self.rfs[worker],
+                    **params,
+                    workers=[worker]
+                )
+            )
+        wait_and_raise_from_futures(model_params)
+        return self
+
+    def _print_summary(self):
+        """
+        Print the summary of the forest used to train and test the model.
+        """
+        futures = list()
+        for n, w in enumerate(self.workers):
+            futures.append(
+                self.client.submit(
+                    _print_summary_func,
+                    self.rfs[w],
+                    workers=[w],
+                )
+            )
+
+        wait_and_raise_from_futures(futures)
+        return self
+
+
+def _func_fit(model, input_data, convert_dtype):
+    X = concatenate([item[0] for item in input_data])

```

Comment:
Not a note for this PR, but this is doing a full copy of the input memory to consolidate the trees, right? @JohnZed, do you think it's possible for the c++ API to accept an "array of arrays" for training the trees? I'm mostly asking because I know they are going through a big refactor. IIRC, Vishal had written the original code to reuse the training data in place. If that's still the case, this shouldn't be too hard of a fix.

Reply:
Good point. Will reach out to them and start discussion on this.